### DISCUSSION TITLE
What are your long term goals?

### DISCUSSION THREAD

**Discussion by teenriot** [daniel-rusu/pods4k#8] [General]
The first sentence of the readme gives room for further goals than the current 'Immutable Arrays'.

> A collection of performance-oriented data structures providing performance and efficiency

Do you plan or like to or think about developing other optimized data structures, e.g Sets once the 'Immutable Arrays' are stable?

What i can imagine is a [Unrolled linked list](https://en.wikipedia.org/wiki/Unrolled_linked_list) which has better memory footprint than ArrayList and LinkedList and unifies the advantages of ArrayList and Linkedlist while being only affected a bit by both disadvantages. If the 'promises' are true, it would be the perfect counterpart on the mutable side, after having optimized the immutable side. I always wondered why there is no standard implementation in Kotlin nor Java. There must be a reason, but i cant figure out what it is. I'm about to develop an implementation, but i will never be as good as you would be.

**Comment by daniel-rusu** [comment]
Thanks for asking!  Yeah, I named this library `Performance Oriented Data Structures for Kotlin` to keep the door open for adding additional data structures in the future.

However, the bar for what gets added will be quite high as there are thousands of promising data structures.  So I want to make sure that I don't spend time implementing and maintaining data structures that won't be widely used.  Essentially, I want to maximize the overall impact on the Kotlin community and I also want to make sure that this library remains sustainable.  I might also include data structures that I have a direct need for in my other projects such as the computer algebra system that I'm building.

Unrolled linked lists are an interesting compromise.  However, sacrificing random-access performance is probably too large of an impact for most performance use-cases, so that's probably why they aren't commonly implemented.  I would prefer to provide solutions that are better than current alternatives in the majority of use-cases.

After Immutable Arrays, I'm considering adding a specialized Set data structure that can perform most complex operations in sub-linear time, such as performing set intersection.  I created a preliminary version of this specialized set data structure in my computer algebra system to efficiently narrow the pattern-matching possibilities and avoid a combinatorial explosion when possible.

About being better, never say never as with enough time and effort, you'll be great at whatever you set your mind to.

**Reply by teenriot** [reply]
Regarding random-access performance for Unrolled Linked list:
My impression from what i read is that the performance sacrifice only a little bit compared to ArrayList. Let's say its below 5 Percent in average than that's not a big deal and totally worth the performance gain for "random access"-mutation. End users are responsible for using the right collections for the right use case. If someone suffers for that 5 Percent random access read, don't use any List at all, use ImmutableArray :) .

Regarding other libraries:
I only know of Eclipse Collections as already maintained library, not making use of Kotlin features as pure Java library. The rest are not maintained any more, for those i know. There are old libs that are still working but they are not adjusted to newer Java versions anymore.
Did you noticed that: https://youtrack.jetbrains.com/issue/KT-55582

**Comment by daniel-rusu** [comment]
> Regarding random-access performance for Unrolled Linked list:
> My impression from what i read is that the performance sacrifice only a little bit compared to ArrayList. Let's say its below 5 Percent in average

It seems impossible for the random-access performance impact to be only 5%, or any fixed percentage, because ArrayList has constant-time access whereas Unrolled Linked List makes random access a linear-time operation.  Unrolled Linked Lists would get slower and slower as they grow since they require at least `N / NodeSize` memory hops to access the `Nth` element.  Memory hops are one of the most expensive operations as each hop could take several hundred CPU cycles depending on where the node is located in memory.

> Did you noticed that: https://youtrack.jetbrains.com/issue/KT-55582

Ooh, thanks for that link!  I added a comment recommending Immutable Arrays.

**Reply by teenriot** [reply]
I think there are ways to get the node for a requested index fast without hopping through memory. However i will do experiments when i have time and let you know if i think i found an algorithm. But for now i'm experimenting for faster ways to get the distinct() operations done. The good old HashSet is not the fastest approach.
